rust/kernel/cpu.rs

```diff
@@ -0,0 +1,127 @@
+///
+/// Cpu values can be safely passed between threads (and across physical CPUs) as they are
+/// merely an integer index representing a CPU. All actual interaction with CPUs is late-
+/// bound.
+#[derive(Copy, Clone)]
+pub struct Cpu(i32);
+
+struct FnCall<F, R>
+where
+    F: FnOnce() -> R,
+{
+    out: Option<R>,
+    func: Option<F>,
+}
+
+unsafe extern "C" fn call_fn<F, R>(fn_box: *mut core::ffi::c_void)
+where
+    F: FnOnce() -> R,
+{
+    // SAFETY: The casting here is the invesrse of the cast to *mut c_void that occurs in Cpu::call
+    // below. That call waits for this call to complete ensuring fn_call is live for the duration.
+    let fn_call = unsafe { &mut *(fn_box as *mut FnCall<F, R>) };
+    let func = fn_call.func.take().unwrap();
+    fn_call.out = Some(func());
+}
+
+impl Cpu {
+    /// Constructs a new instance referring to a specific processor by id.
+    ///
+    /// This function checks that the passed id is a valid _potential_ processor id, but does
+    /// not check that the processor is actually present or online. Manipulation of absent
+    /// or offline processors will fail with reasonable errors returned.
+    pub fn from_id(id: i32) -> Result<Cpu> {
+        if id >= Self::num_possible() {
+            return Err(E2BIG);
+        } else if id < 0 {
+            return Err(EINVAL);
+        }
+        Ok(Cpu(id))
+    }
+
+    /// Returns the number of processors that are possible (including onlining and hotplug).
+    pub fn num_possible() -> i32 {
+        // SAFETY: The number of possible CPUs is fixed early at boot and does not change in
+        // response to processors coming online, offline, or hotplug events.
+        unsafe { bindings::nr_cpu_ids as i32 }
+    }
+
+    /// Lock the running thread to the current CPU and prevent preemption.
+    pub fn lock_current() -> Guard {
+        Guard::new()
+    }
+
+    /// Get the processor id of this CPU.
+    pub fn id(&self) -> i32 {
+        self.0
+    }
+
+    /// Call a function on the CPU identified by `self`.
+    pub fn call<R, F>(&self, func: F) -> Result<R>
+    where
+        F: FnOnce() -> R + Send,
+        R: Send,
+    {
+        let mut fn_call = FnCall {
+            out: None,
+            func: Some(func),
+        };
+        // SAFETY: smp_call_function_single is invoked with wait == 1 (true) ensuring that fn_call
+        // outlives this FFI invocation. func's returned value is moved into fn_call.out and then
+        // returned here. The Send type constraint ensures func and its returned value can safely
+        // be moved between threads/CPUs.
+        unsafe {
+            let result = bindings::smp_call_function_single(
+                self.0,
+                Some(call_fn::<F, R>),
+                (&mut fn_call as *mut FnCall<F, R>).cast(),
+                1,
+            );
+            if result < 0 {
+                return Err(Error::from_kernel_errno(result));
+            }
+        }
```

Comment:
```suggestion
       let result = unsafe {
            bindings::smp_call_function_single(
                self.0,
                Some(call_fn::<F, R>),
                (&mut fn_call as *mut FnCall<F, R>).cast(),
                1,
            )
        };
        if result < 0 {
                return Err(Error::from_kernel_errno(result));
        }
```

To minimize the scope of `unsafe`. Typing code in github is hard, so you may want to reformat that with rustfmt ;-)

Reply:
I went with @wedsonaf's suggestion and used to_result.